test/BootcampFactory.t.sol

```diff
@@ -0,0 +1,235 @@
+        );
+
+        factory.grantRole(MANAGER, address(0));
+    }
+
+    function test_AdminCanNotGrantNonExistentRole() public {
+        vm.prank(admin);
+        vm.expectRevert(abi.encodeWithSelector(
+            IBootcampFactoryErrors.BootcampFactory__UpdateNonExistentRole.selector, 
+            INVALID_ROLE
+        ));
+
+        factory.grantRole(INVALID_ROLE, manager);
+    }
+
+
+
+    /*//////////////////////////////////////////////////
+                REVOKEROLE TESTS
+    /////////////////////////////////////////////////*/
+    function test_AdminSuccessfullyRevokeManagerRoleFromUser() public adminGrantManagerRole {
+        factory.revokeRole(MANAGER, manager);
+        
+        assertFalse(factory.hasRole(MANAGER, manager));
+    }
+    
+     function test_AdminSuccessfullyRevokeAdminRoleFromUser() public adminGrantManagerRole {
+        factory.revokeRole(ADMIN, alice);
+
+        assertFalse(factory.hasRole(ADMIN, alice));
+    }
+
+    function test_OnlyAdminCanRevokeTheRoles() public adminGrantManagerRole {
+        vm.startPrank(alice);
+        vm.expectRevert(abi.encodeWithSelector(
+            IAccessControl.AccessControlUnauthorizedAccount.selector,
+            alice,
+            ADMIN
+        ));
+
+        factory.revokeRole(MANAGER, manager);
+    }
+
+    function test_AccountLosingARoleCanNotBeAddressZero() public adminGrantManagerRole {
+        vm.expectRevert(
+            IBootcampFactoryErrors.BootcampFactory__CanNotUpdateRoleForZeroAddress.selector
+        );
+
+        factory.revokeRole(MANAGER, address(0));
+    }
+    
+    function test_AdminCanNotRevokeNonExistentRole() public adminGrantManagerRole {
+        vm.expectRevert(abi.encodeWithSelector(
+            IBootcampFactoryErrors.BootcampFactory__UpdateNonExistentRole.selector, 
+            INVALID_ROLE
+        ));
+
+        factory.revokeRole(INVALID_ROLE, manager);
+    }
+
+
+    
+    /*//////////////////////////////////////////////////
+                CREATEBOOTCAMP TESTS
+    /////////////////////////////////////////////////*/
+    function test_ManagerSuccessfullyCreateNewBootcamp() public adminGrantManagerRole {
+        vm.startPrank(manager);
+        factory.createBootcamp(
+            networkConfig.depositAmount,
+            networkConfig.depositToken,
+            networkConfig.bootcampStartTime
+        );
+        uint256 bootcampId = factory.totalBootcampAmount();
+        (
+            uint256 id, 
+            uint256 depositAmount, 
+            address depositToken, 
+            address bootcampAddress
+        ) = factory.bootcamps(bootcampId);
+        
+        assertEq(id, 1);
+        assertEq(depositAmount, networkConfig.depositAmount);
+        assertEq(depositToken, networkConfig.depositToken);
+        assertTrue(bootcampAddress != address(0));
+    }
+
+    function test_EventIsEmittedOnceBootcampIsCreated() public adminGrantManagerRole {
```

Comment:
Thanks, it is actually important to make sure this is working properly if we don't use the counter !

Reply:
Of course, I'll update tests once update a contract)